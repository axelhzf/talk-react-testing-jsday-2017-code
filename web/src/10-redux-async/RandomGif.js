import React from 'react';
import { applyMiddleware, combineReducers, compose, createStore } from 'redux';
import thunk from 'redux-thunk';
import { connect } from 'react-redux';
import axios from 'axios';

function gifReducer(state = { state: undefined }, action) {
  switch (action.type) {
    case 'FETCH_GIF':
      return { state: 'loading' };
    case 'FETCH_GIF_SUCCESS':
      return { state: 'success', url: action.url };
    case 'FETCH_GIF_ERROR':
      return { state: 'error' };
    default:
      return state;
  }
}

export const reducer = combineReducers({
  gif: gifReducer
});

const fetchGif = () => async dispatch => {
  dispatch({ type: 'FETCH_GIF' });
  try {
    const response = await axios({
      method: 'POST',
      url: 'https://www.graphqlhub.com/graphql',
      data: { query: getRandomGifQuery }
    });
    const url = response.data.data.giphy.random.images.original.url;
    dispatch({ type: 'FETCH_GIF_SUCCESS', url });
  } catch (e) {
    dispatch({ type: 'FETCH_GIF_ERROR' });
  }
};

export const createDefaultStore = () => {
  const composeEnhancers =
    typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
      ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({
          name: 'redux-async'
        })
      : compose;
  const enhancer = composeEnhancers(applyMiddleware(thunk));
  const store = createStore(reducer, enhancer);
  return store;
};

class RandomGif extends React.Component {
  componentDidMount() {
    this.props.fetchGif();
  }

  render() {
    const { gif, fetchGif } = this.props;
    if (gif.state === undefined) return null;
    if (gif.state === 'loading') return <div>Loading...</div>;
    if (gif.state === 'error') return <div>Error</div>;
    return (
      <div>
        <img src={gif.url} />
        <button onClick={fetchGif}>Random</button>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  gif: state.gif
});

export const RandomGifContainer = connect(mapStateToProps, { fetchGif })(
  RandomGif
);

const getRandomGifQuery = `
{
  giphy {
    random(tag: "cool") {
      images {
        original {
          url
        }
      }
    }
  }
}
`;
