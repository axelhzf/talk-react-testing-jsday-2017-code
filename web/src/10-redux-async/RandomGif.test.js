import React from 'react';
import { RandomGifContainer, reducer } from './RandomGif';
import { applyMiddleware, createStore } from 'redux';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import thunk from 'redux-thunk';
import _ from 'lodash';

const mock = new MockAdapter(axios);

describe('10-redux-async', () => {

  beforeEach(() => {
    mock.reset();
  });

  it('should render on fetch success', async () => {
    const observer = new ReduxObserver();
    const store = createStore(reducer, applyMiddleware(thunk, observer.middleWare));

    mock.onAny().reply(config => {
      const body = {
        data: {
          giphy: {
            random: {
              images: {
                original: {
                  url: 'https://media2.giphy.com/media/2HONNTJbRhzKE/giphy.gif'
                }
              }
            }
          }
        }
      };
      return [200, body];
    });

    const wrapper = mount(
      <Provider store={store}>
        <RandomGifContainer/>
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();

    await observer.waitForActionType('FETCH_GIF_SUCCESS');
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render on fetch error', async () => {
    const observer = new ReduxObserver();
    const store = createStore(reducer, applyMiddleware(thunk, observer.middleWare));

    mock.onAny().reply(500);

    const wrapper = mount(
      <Provider store={store}>
        <RandomGifContainer/>
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();

    await observer.waitForActionType('FETCH_GIF_ERROR');
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
});

class ReduxObserver {

  snapshots = [];
  pendingPromises = {}

  middleWare = store => next => action => {
    const state = store.getState();
    next(action);
    const nextState = store.getState();

    this.snapshots.push({ state, action, nextState });

    const pendingPromise = this.pendingPromises[action.type];
    if (pendingPromise) pendingPromise();

    return nextState;
  }

  waitForActionType (actionType) {
    return new Promise((resolve) => {
      const action = _.find(this.snapshots, ({ action }) => action.type === actionType);
      if (action) return resolve();
      this.pendingPromises[actionType] = resolve;
    });
  }

}
