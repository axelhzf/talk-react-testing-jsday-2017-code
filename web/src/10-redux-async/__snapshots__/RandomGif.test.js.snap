// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`10-redux-async should render on fetch error 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(RandomGif)>
    <RandomGif
      fetchGif={[Function]}
      gif={
        Object {
          "state": "loading",
        }
      }
    >
      <div>
        Loading...
      </div>
    </RandomGif>
  </Connect(RandomGif)>
</Provider>
`;

exports[`10-redux-async should render on fetch error 2`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(RandomGif)>
    <RandomGif
      fetchGif={[Function]}
      gif={
        Object {
          "state": "error",
        }
      }
    >
      <div>
        Error
      </div>
    </RandomGif>
  </Connect(RandomGif)>
</Provider>
`;

exports[`10-redux-async should render on fetch success 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(RandomGif)>
    <RandomGif
      fetchGif={[Function]}
      gif={
        Object {
          "state": "loading",
        }
      }
    >
      <div>
        Loading...
      </div>
    </RandomGif>
  </Connect(RandomGif)>
</Provider>
`;

exports[`10-redux-async should render on fetch success 2`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(RandomGif)>
    <RandomGif
      fetchGif={[Function]}
      gif={
        Object {
          "state": "success",
          "url": "https://media2.giphy.com/media/2HONNTJbRhzKE/giphy.gif",
        }
      }
    >
      <div>
        <img
          src="https://media2.giphy.com/media/2HONNTJbRhzKE/giphy.gif"
        />
        <button
          onClick={[Function]}
        >
          Random
        </button>
      </div>
    </RandomGif>
  </Connect(RandomGif)>
</Provider>
`;
